substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -1185,8 +1189,12 @@ impl<T: Config> Pallet<T> {
         // If the contract is in delete state, remove all associated storage
         if matches!(contract.state, types::ContractState::Deleted(_)) {
             Self::remove_contract(contract.contract_id);
+            return Ok(().into());
         }
```

Comment:
In the `bill_contract()` context, the only way to have a contract in `Deleted` state is just after the `handle_grace()`
So i guess this block can be moved above
With this you can get rid of the `if` at line 1157

Reply:
Well, the contract is also moved to state `Deleted` if the user cancels it: https://github.com/threefoldtech/tfchain/blob/af0a2d546d11796a311ddbae52f9610227edd932/substrate-node/pallets/pallet-smart-contract/src/lib.rs#L954

Reply:
Btw, I don't understand why there is a `Deleted` state for contract because I don't see something specific done with this info 
Just removing the contract would not be fine?
or I miss something

Reply:
> Well, the contract is also moved to state `Deleted` if the user cancels it:
> 
> https://github.com/threefoldtech/tfchain/blob/af0a2d546d11796a311ddbae52f9610227edd932/substrate-node/pallets/pallet-smart-contract/src/lib.rs#L954

Ok, I see
So even when `Deleted` it is billed and need to reach the end of `bill_contract()`